vignettes/creating-custom-modules.Rmd

```diff
@@ -12,173 +12,254 @@ vignette: >
-      plot_code_q()[["p"]]
+    # Render the histogram from the updated teal_data object
+    output$histogram_plot <- shiny::renderPlot({
+      result()[["my_plot"]]  # Access and render the plot stored in `new_data`
     })
 
-    # view code
-    output$code <- renderText({
-      get_code(plot_code_q())
+    # Reactive expression to get the generated code for the plot
+    output$plot_code <- shiny::renderText({
+      teal.code::get_code(result())  # Retrieve and display the code for the updated `teal_data` object
     })
   })
 }
+```
+
+Let's review what we've done so far:
+
+1. **Dynamic Variable Injection with `within()`**:
+    - `input_dataset = as.name(input$dataset)` passes the dataset name dynamically as `input_dataset`.
+    - `input_vars = input$variable` passes the selected variable name directly as `input_vars`.
+    - Inside `within()`, `my_plot` uses these injected variables to dynamically generate the histogram plot.
 
-# function that creates module instance to use in `teal` app
-tm_histogram_example <- function(label) {
-  module(
+2. **Rendering the Plot**:
+    - `output$histogram_plot` uses `renderPlot()` to display the plot stored in `new_data` by referencing `result()[["my_plot"]]`.
+
+3. **Plot Code Display**:
+    - The `output$plot_code` render function displays the dynamically generated code using `teal.code::get_code(result())`, allowing users to see the exact code used to generate the plot reactively.
+
+## Creating the Custom `teal` Module Function
+
+The `teal::module()` function allows you to encapsulate your UI and server logic into a `teal` module, making it reusable and ready to integrate into any `teal` application.
+
+By setting `datanames = "all"`, you give the module access to all datasets specified in the `teal_data` object.
+
+```r
+# Custom histogram module creation
+create_histogram_module <- function(label = "Histogram Module") {
+  teal::module(
     label = label,
-    server = srv_histogram_example,
-    ui = ui_histogram_example,
+    ui = histogram_module_ui,
+    server = histogram_module_server,
     datanames = "all"
   )
 }
 ```
 
-This module is ready to be used in a `teal` app.
+## Integrating the Custom `teal` Module into a `teal` App
+
+With the custom `teal` module set up, it can now be integrated into a `teal` app. 
+We’ll use `init()` from `teal` to specify the datasets and modules used in the app, then run the app to test the newly created module.
+
+```r
+library(teal)
+
+# Define datasets in `teal_data`
+data_obj <- teal_data(
+  iris = iris,
+  mtcars = mtcars
+)
 
-```{r}
+# Initialize the teal app
 app <- init(
-  data = teal_data(IRIS = iris, NPK = npk),
-  modules = tm_histogram_example(label = "Histogram Module"),
-  header = "Simple app with custom histogram module"
+  data = data_obj,
+  modules = modules(create_histogram_module())
 )
 
-if (interactive()) {
-  shinyApp(app$ui, app$server)
-}
+# Run the app
+shiny::shinyApp(ui = app$ui, server = app$server)
 ```
```

Comment:
WDYT about adding a Shinylive compontent in here - right after the final app?

More general issue: https://github.com/insightsengineering/teal/issues/1404. We can do it now in this PR or wait for this.

Reply:
I did attempt to add the link as you suggested but it's not working because we haven't release the updated version of teal.data (using `names` instead of `datanames` when pulling the names of data objects in teal_data object). I think we should wait.